### ISSUE TITLE
add return value in the variable window

### ISSUE THREAD

**Issue by gitlsl** [0xd4d/dnSpy#798]
Sometimes you may encounter a very long and complex expression.
so it is very helpful if record the method return value in the expression or Split a complex expression into multiple simple ones.

**Comment by rolandh** [comment]
How do I enable this? I tried adding $ReturnValue to the watch window but it does not appear to work.

**Comment by 0xd4d** [comment]
It only works if you're debugging .NET Framework 4.5.1+ or .NET Core 1.x+ programs and if it's a supported return value (eg. not value types unless they're primitives) https://docs.microsoft.com/dotnet/framework/unmanaged-api/debugging/icordebugcode3-getreturnvalueliveoffset-method

**Comment by 0xd4d** [comment]
The local window should also show it so you don't need to use the watch window.

**Comment by rolandh** [comment]
Yeah it is a value type almost always in my case so that explains it. Anyway to to easily watch these values? Is it possible to add an expression to the watch window?

**Comment by 0xd4d** [comment]
`SomeMethod(arg1,arg2)` but that has side effects.

**Comment by 0xd4d** [comment]
BTW, if you want the debugger to support struct return values, you have to file a [CoreCLR issue](https://github.com/dotnet/coreclr/issues) since all .NET debuggers use the same `ICorDebug` API

**Comment by rolandh** [comment]
Thanks

**Comment by rolandh** [comment]
Is there any $ keywords to show the args passed to the method? I have a static method that has an integer passed to it, however I cannot view what the number is in the locals or watch window. Is this because it is an integer? If so is there any way to print this value to console? I have tried the breakpoint however if I print {value} to the console the value is wrong.

**Comment by 0xd4d** [comment]
You should be able to just use the name of the arg as you see in the decompiled code. If that doesn't work, you should see some error message shown in the watch window or output window if you use a tracepoint. What's the error message?

**Comment by rolandh** [comment]
In the locals window the argument simply doesn't show up at all.

Here is the method spec
public static string blah(int A_0)

If I manually type A_0 into the watch window I get: 
error CS0103: The name 'A_0' does not exist in the current context

If I type "value" into the watch window I get a value which is about 25% different to A_0. I tried to add a breakpoint and log message to print A_0

The method is called via a delegate, I wonder if this confuses the debugger. Is there any way to watch the local variables in the memory window easily? I'm not sure how to find the base address to find the values.

Any other debugging I could do to figure out why it doesn't work?

**Comment by 0xd4d** [comment]
`A_0` probably means that there's no parameter name in the metadata, could be the reason Roslyn complains. You can try to enable 'show raw locals' in the debugger settings.

**Comment by rolandh** [comment]
Thanks that worked. A_0 still didn't match but I figured out why. The "light weight function" was an ILGenerator emit which modified the called value.

**Comment by rolandh** [comment]
Different topic but is there anyway to execute a script when a breakpoint occurs? I know I could just write my own program however sometimes it would be nice to write 2 or 3 lines of code that executes when a breakpoint is hit to print out say callerClass.Methods(x => x.signature == something).MDToken etc

**Comment by 0xd4d** [comment]
It's not possible, open a new issue if it's a useful feature.

**Comment by 0xd4d** [comment]
Tracepoints can be used too to execute some code.
